File: clients/web/src/application/MessagePipeline.ts
import type {
  AttachmentCapability,
  AuthSession,
  DeviceSyncPayload,
  GatewayAcknowledgement,
  GatewaySyncResponse,
  LocalMessage,
  OutboxRecord,
  WireMessage,
} from "../domain/contracts";
import { MessagePayloadCodec } from "../domain/MessagePayloadCodec";
import { decodeText } from "../domain/encoding";
import { WasmCrypto } from "../crypto/WasmCrypto";
import { ApiClient } from "../infrastructure/ApiClient";
import { CredentialManager } from "../infrastructure/CredentialManager";
import { GatewayClient, GatewayError } from "../infrastructure/GatewayClient";
import { LocalRepository } from "../infrastructure/LocalRepository";
import { IdentityTrustService } from "./IdentityTrustService";

export interface PipelineUpdate {
  messages: LocalMessage[];
  error: string | null;
}

export class MessagePipeline {
  private queue: Promise<void> = Promise.resolve();

  constructor(
    private readonly api: ApiClient,
    private readonly credentials: CredentialManager,
    private readonly repository: LocalRepository,
    private readonly crypto: WasmCrypto,
    private readonly gateway: GatewayClient,
    private readonly identityTrust: IdentityTrustService,
    private readonly payloads: MessagePayloadCodec,
  ) {}

  activate(session: AuthSession, activePeer: string | null): Promise<PipelineUpdate> {
    return this.serialized(async () => {
      let error: string | null = null;
      for (const inbox of await this.repository.pendingInbox(session.device_id)) {
        try {
          await this.processIncoming(session, inbox.message, activePeer);
        } catch (failure) {
          error = this.errorMessage(failure);
        }
      }
      return { messages: await this.repository.messages(session.device_id), error };
    });
  }

  enqueue(
    recipientUsername: string,
    body: string,
    session: AuthSession,
  ): Promise<PipelineUpdate> {
    return this.enqueuePayload(recipientUsername, body, null, session);
  }

  enqueueAttachment(
    recipientUsername: string,
    capability: AttachmentCapability,
    session: AuthSession,
  ): Promise<PipelineUpdate> {
    return this.enqueuePayload(recipientUsername, capability.filename, capability, session);
  }

  private enqueuePayload(
    recipientUsername: string,
    body: string,
    attachment: AttachmentCapability | null,
    session: AuthSession,
  ): Promise<PipelineUpdate> {
    return this.serialized(async () => {
      const directory = await this.credentials.authorized((token) =>
        this.api.keyBundles(token, recipientUsername)
      );
      if (directory.devices.length === 0) {
        throw new Error("Recipient has no active devices");
      }
      const trust = await this.identityTrust.verify(
        session.device_id,
        directory.username,
        directory.devices,
      );
      const messageId = crypto.randomUUID();
      const prepared = await this.crypto.prepareEncryptionBytes(
        session.device_id,
        directory.devices,
        attachment
          ? this.payloads.encodeAttachment(attachment)
          : this.payloads.encodeText(body),
      );
      const createdAt = new Date().toISOString();
      const message: LocalMessage = {
        id: messageId,
        recipient_user_id: directory.user_id,
        peer_username: directory.username,
        direction: "outgoing",
        body,
        created_at: createdAt,
        sender_device_id: session.device_id,
        recipient_device_ids: prepared.envelopes.map((envelope) => envelope.recipient_device_id),
        delivery_state: "sending",
        failure_reason: null,
        attachment,
      };
      const outbox: OutboxRecord = {
        id: messageId,
        local_device_id: session.device_id,
        recipient_user_id: directory.user_id,
        recipient_username: directory.username,
        plaintext: body,
        envelopes: prepared.envelopes,
        created_at: createdAt,
        state: "sending",
        attempts: 0,
        retry_at: null,
        failure_reason: null,
        device_set_changed: false,
        attachment: attachment ?? undefined,
      };
      await this.repository.stageOutgoing({
        outbox,
        message,
        sessions: prepared.sessions,
        trust,
      });
      const error = await this.deliver(outbox);
      if (!error) {
        await this.enqueueSelfSync(session, {
          version: 1,
          kind: "outgoing_message",
          logical_message_id: messageId,
          occurred_at: Date.now(),
          outgoing_message: {
            recipient_user_id: directory.user_id,
            recipient_username: directory.username,
            body,
            sent_at: Date.parse(createdAt),
            attachment: attachment ?? undefined,
          },
        });
      }
      return { messages: await this.repository.messages(session.device_id), error };
    });
  }

  ingest(
    session: AuthSession,
    message: WireMessage,
    activePeer: string | null,
  ): Promise<PipelineUpdate> {
    return this.serialized(async () => {
      let error: string | null = null;
      try {
        await this.processIncoming(session, message, activePeer);
        await this.flushAcknowledgements(session.device_id);
      } catch (failure) {
        error = this.errorMessage(failure);
      }
      return { messages: await this.repository.messages(session.device_id), error };
    });
  }

  ingestSync(
    session: AuthSession,
    response: GatewaySyncResponse,
    activePeer: string | null,
  ): Promise<PipelineUpdate> {
    return this.serialized(async () => {
      let error: string | null = null;
      for (const message of response.messages) {
        try {
          await this.processIncoming(session, message, activePeer);
        } catch (failure) {
          error = this.errorMessage(failure);
        }
      }
      await this.repository.saveCursor(session.device_id, response.next_cursor);
      await this.flushAcknowledgements(session.device_id);
      return { messages: await this.repository.messages(session.device_id), error };
    });
  }

  synchronize(session: AuthSession, activePeer: string | null): Promise<PipelineUpdate> {
    return this.serialized(async () => {
      let cursor = await this.repository.cursor(session.device_id);
      let error: string | null = null;
      for (;;) {
        const response = await this.gateway.sync(cursor, 100);
        for (const message of response.messages) {
          try {
            await this.processIncoming(session, message, activePeer);
          } catch (failure) {
            error = this.errorMessage(failure);
          }
        }
        cursor = response.next_cursor;
        await this.repository.saveCursor(session.device_id, cursor);
        await this.flushAcknowledgements(session.device_id);
        if (response.messages.length < 100) {
          break;
        }
      }
      return { messages: await this.repository.messages(session.device_id), error };
    });
  }

  flushOutbox(session: AuthSession): Promise<PipelineUpdate> {
    return this.serialized(async () => {
      let error: string | null = null;
      for (const record of await this.repository.readyOutbox(session.device_id)) {
        error = await this.deliver(record) ?? error;
      }
      return { messages: await this.repository.messages(session.device_id), error };
    });
  }

  retry(session: AuthSession, messageId: string): Promise<PipelineUpdate> {
    return this.serialized(async () => {
      const record = await this.repository.outbox(session.device_id, messageId);
      if (!record) {
        throw new Error("Outbox message is unavailable");
      }
      if (record.device_set_changed) {
        await this.reprepare(session, record);
      } else {
        await this.repository.retryOutbox(session.device_id, messageId);
      }
      const ready = await this.repository.outbox(session.device_id, messageId);
      const error = ready ? await this.deliver(ready) : "Outbox message is unavailable";
      return { messages: await this.repository.messages(session.device_id), error };
    });
  }

  synchronizeReadState(session: AuthSession, peerUsername: string): Promise<void> {
    return this.serialized(async () => {
      await this.enqueueSelfSync(session, {
        version: 1,
        kind: "read_state",
        logical_message_id: crypto.randomUUID(),
        occurred_at: Date.now(),
        read_state: { peer_username: peerUsername, read_at: Date.now() },
      });
    });
  }

  private async processIncoming(
    session: AuthSession,
    message: WireMessage,
    activePeer: string | null,
  ): Promise<GatewayAcknowledgement> {
    if (
      message.recipient_user_id !== session.user_id ||
      message.recipient_device_id !== session.device_id
    ) {
      throw new Error("Gateway delivered a message for another device");
    }
    await this.repository.saveInbox(session.device_id, message);
    if (await this.repository.hasMessage(session.device_id, message.id)) {
      return this.repository.commitDuplicate(session.device_id, message);
    }
    const prepared = await this.crypto.prepareDecryption(session.device_id, message);
    const payload = this.payloads.decode(prepared.plaintext);
    if (payload?.kind === "device_sync") {
      if (
        message.sender_user_id !== session.user_id ||
        message.sender_username.toLowerCase() !== session.username.toLowerCase()
      ) {
        throw new Error("Device sync was not sent by this account");
      }
      return this.repository.commitDeviceSync(
        session.device_id,
        prepared.remoteDeviceId,
        prepared.session,
        prepared.preKeyState,
        message,
        payload.device_sync,
      );
    }
    const body = payload?.kind === "text"
      ? payload.text
      : payload?.kind === "attachment"
        ? payload.attachment.filename
        : decodeText(prepared.plaintext);
    return this.repository.commitIncoming({
      localDeviceId: session.device_id,
      remoteDeviceId: prepared.remoteDeviceId,
      session: prepared.session,
      preKeyState: prepared.preKeyState,
      message: {
        ...prepared.message,
        body,
        attachment: payload?.kind === "attachment" ? payload.attachment : null,
      },
      wireMessage: message,
      isRead: activePeer?.toLowerCase() === message.sender_username.toLowerCase(),
    });
  }

  private async flushAcknowledgements(deviceId: string): Promise<void> {
    const acknowledgements = await this.repository.readyAcknowledgements(deviceId);
    if (acknowledgements.length === 0) {
      return;
    }
    await this.gateway.acknowledge(acknowledgements);
    await this.repository.markAcknowledged(deviceId, acknowledgements);
  }

  private async deliver(record: OutboxRecord): Promise<string | null> {
    try {
      await this.gateway.send(record.id, record.recipient_user_id, record.envelopes);
      await this.repository.markOutboxSent(record.local_device_id, record.id);
      return null;
    } catch (failure) {
      const reason = this.errorMessage(failure);
      if (
        failure instanceof GatewayError &&
        ["offline", "disconnected", "timeout"].includes(failure.code)
      ) {
        await this.repository.deferOutbox(record.local_device_id, record.id, reason);
        return reason;
      }
      await this.repository.failOutbox(
        record.local_device_id,
        record.id,
        reason,
        failure instanceof GatewayError && failure.deviceSetChanged,
      );
      return reason;
    }
  }

  private async reprepare(session: AuthSession, record: OutboxRecord): Promise<void> {
    const directory = await this.credentials.authorized((token) =>
      record.device_sync
        ? this.api.ownKeyBundles(token)
        : this.api.keyBundles(token, record.recipient_username)
    );
    const trust = await this.identityTrust.verify(
      session.device_id,
      directory.username,
      directory.devices,
    );
    const payload = record.device_sync
      ? this.payloads.encodeDeviceSync(record.device_sync)
      : record.attachment
        ? this.payloads.encodeAttachment(record.attachment)
        : this.payloads.encodeText(record.plaintext);
    const prepared = await this.crypto.prepareEncryptionBytes(
      session.device_id,
      directory.devices,
      payload,
    );
    await this.repository.replaceOutbox(
      session.device_id,
      record.id,
      directory.user_id,
      prepared.envelopes,
      prepared.sessions,
      trust,
    );
  }

  private async enqueueSelfSync(
    session: AuthSession,
    payload: DeviceSyncPayload,
  ): Promise<void> {
    const directory = await this.credentials.authorized((token) => this.api.ownKeyBundles(token));
    if (directory.devices.length === 0) {
      return;
    }
    const trust = await this.identityTrust.verify(
      session.device_id,
      session.username,
      directory.devices,
    );
    const prepared = await this.crypto.prepareEncryptionBytes(
      session.device_id,
      directory.devices,
      this.payloads.encodeDeviceSync(payload),
    );
    const transportId = crypto.randomUUID();
    const outbox: OutboxRecord = {
      id: transportId,
      local_device_id: session.device_id,
      recipient_user_id: session.user_id,
      recipient_username: session.username,
      plaintext: JSON.stringify(payload),
      envelopes: prepared.envelopes,
      created_at: new Date().toISOString(),
      state: "sending",
      attempts: 0,
      retry_at: null,
      failure_reason: null,
      device_set_changed: false,
      visible: false,
      device_sync: payload,
    };
    await this.repository.stageDeviceSyncOutgoing(outbox, prepared.sessions, trust);
    await this.deliver(outbox);
  }

  private serialized<Result>(operation: () => Promise<Result>): Promise<Result> {
    const result = this.queue.then(operation, operation);
    this.queue = result.then(() => undefined, () => undefined);
    return result;
  }

  private errorMessage(error: unknown): string {
    return error instanceof Error ? error.message : "Messaging operation failed";
  }
}

File: clients/apple/Knot/Knot/Services/MessagePipeline.swift
import Foundation

actor MessagePipeline {
  private let api: APIClient
  private let crypto: CryptoService
  private let store: EncryptedMessageStore
  private let maximumAttachmentCiphertextSize: Int64

  init(
    api: APIClient,
    crypto: CryptoService,
    store: EncryptedMessageStore,
    maximumAttachmentCiphertextSize: Int64
  ) {
    self.api = api
    self.crypto = crypto
    self.store = store
    self.maximumAttachmentCiphertextSize = maximumAttachmentCiphertextSize
  }

  func activate(session: AuthSession, selectedConversationID: String?) async throws
    -> PipelineUpdate
  {
    try await store.activate(userID: session.userID, deviceID: session.deviceID)
    var acknowledgements = try await recoverMutations()
    let pending = try await store.pendingInbox()
    var processingError: String?
    for message in pending {
      do {
        if let acknowledgement = try await processIncoming(
          message,
          session: session,
          selectedConversationID: selectedConversationID
        ) {
          acknowledgements.append(acknowledgement)
        }
      } catch {
        processingError = error.localizedDescription
      }
    }
    acknowledgements.append(contentsOf: try await store.readyAcknowledgements())
    acknowledgements = unique(acknowledgements)
    return PipelineUpdate(
      conversations: try await store.conversations(),
      acknowledgements: acknowledgements,
      processingError: processingError
    )
  }

  func deactivate() async {
    await store.close()
  }

  func addConversation(username: String) async throws -> [Conversation] {
    try await store.addConversation(username: username)
    return try await store.conversations()
  }

  func enqueue(
    messageID: String,
    body: String,
    attachment: AttachmentCapability?,
    plaintext: Data,
    conversationID: String,
    recipientUsername: String,
    sentAt: Date,
    session: AuthSession
  ) async throws -> [Conversation] {
    try await store.addConversation(username: recipientUsername)
    try await store.enqueueOutgoingMessage(
      id: messageID,
      conversationID: conversationID,
      recipientUsername: recipientUsername,
      body: body,
      attachment: attachment,
      plaintext: plaintext,
      sentAt: sentAt
    )
    var stagedMutation = false
    do {
      let conversation = try await store.conversation(id: conversationID)
      let directory = try await api.keyDirectory(
        username: conversation.username,
        accessToken: session.accessToken
      )
      guard !directory.devices.isEmpty else {
        throw MessagingError.noRecipientDevices
      }
      try await store.updateConversationDirectory(
        id: conversationID,
        username: directory.username,
        recipientUserID: directory.userID
      )
      let prepared = try await crypto.prepareMessage(
        plaintext: plaintext,
        ownerUsername: session.username,
        username: directory.username,
        devices: directory.devices
      )
      do {
        try await store.stageOutbound(
          messageID: messageID,
          recipientUsername: directory.username,
          recipientUserID: directory.userID,
          envelopes: prepared.envelopes,
          mutation: prepared.mutation
        )
      } catch {
        await crypto.discard(prepared.mutation)
        throw error
      }
      stagedMutation = true
      _ = try await recoverMutations()
      return try await store.conversations()
    } catch {
      if !stagedMutation {
        _ = try? await store.scheduleRetry(
          messageID: messageID,
          error: error.localizedDescription
        )
      }
      throw error
    }
  }

  func ingest(
    messages: [ServerMessage],
    nextCursor: String?,
    advanceCursor: Bool,
    selectedConversationID: String?,
    session: AuthSession
  ) async throws -> PipelineUpdate {
    let validMessages = messages.filter {
      $0.recipientUserID == session.userID && $0.recipientDeviceID == session.deviceID
    }
    let batch = try await store.ingest(
      messages: validMessages,
      nextCursor: nextCursor,
      advanceCursor: advanceCursor
    )
    var acknowledgements = batch.acknowledgements
    var processingError: String?
    for message in batch.pending {
      do {
        if let acknowledgement = try await processIncoming(
          message,
          session: session,
          selectedConversationID: selectedConversationID
        ) {
          acknowledgements.append(acknowledgement)
        }
      } catch {
        processingError = error.localizedDescription
      }
    }
    return PipelineUpdate(
      conversations: try await store.conversations(),
      acknowledgements: unique(acknowledgements),
      processingError: processingError
    )
  }

  func cursor() async throws -> String {
    try await store.cursor()
  }

  func readyOutbox() async throws -> [OutboxRecord] {
    try await store.readyOutbox()
  }

  func markSent(messageID: String) async throws -> [Conversation] {
    try await store.markSent(messageID: messageID)
    return try await store.conversations()
  }

  func scheduleRetry(messageID: String, error: String) async throws -> [Conversation] {
    _ = try await store.scheduleRetry(messageID: messageID, error: error)
    return try await store.conversations()
  }

  func retry(messageID: String) async throws -> [Conversation] {
    try await store.retry(messageID: messageID)
    return try await store.conversations()
  }

  func reprepare(_ record: OutboxRecord, session: AuthSession) async throws -> [Conversation] {
    let directory: UserKeyDirectory
    if record.isDeviceSync {
      directory = try await api.ownKeyDirectory(accessToken: session.accessToken)
    } else {
      directory = try await api.keyDirectory(
        username: record.payload.recipientUsername,
        accessToken: session.accessToken
      )
    }
    guard !directory.devices.isEmpty else {
      throw MessagingError.noRecipientDevices
    }
    let prepared = try await crypto.prepareMessage(
      plaintext: record.payload.plaintext,
      ownerUsername: session.username,
      username: directory.username,
      devices: directory.devices
    )
    do {
      if !record.isDeviceSync {
        try await store.updateConversationDirectory(
          id: record.conversationID,
          username: directory.username,
          recipientUserID: directory.userID
        )
      }
      try await store.reprepareOutbound(
        messageID: record.id,
        recipientUsername: directory.username,
        recipientUserID: directory.userID,
        envelopes: prepared.envelopes,
        mutation: prepared.mutation
      )
    } catch {
      await crypto.discard(prepared.mutation)
      throw error
    }
    _ = try await recoverMutations()
    return try await store.conversations()
  }

  func markAcknowledged(_ acknowledgements: [GatewayAcknowledgement]) async throws {
    try await store.markAcknowledged(acknowledgements)
  }

  func markRead(conversationID: String) async throws -> [Conversation] {
    try await store.markRead(conversationID: conversationID)
    return try await store.conversations()
  }

  func enqueueOutgoingDeviceSync(_ record: OutboxRecord, session: AuthSession) async throws {
    guard !record.isDeviceSync,
      let decoded = try MessagePayloadCodec.decode(
        record.payload.plaintext,
        maximumCiphertextSize: maximumAttachmentCiphertextSize
      )
    else {
      return
    }
    let body: String
    let attachment: AttachmentCapability?
    switch decoded.kind {
    case .text:
      guard let text = decoded.text else {
        throw MessagingError.invalidPlaintext
      }
      body = text
      attachment = nil
    case .attachment:
      guard let capability = decoded.attachment else {
        throw MessagingError.invalidPlaintext
      }
      body = capability.filename
      attachment = capability
    case .deviceSync:
      return
    }
    try await enqueueDeviceSync(
      DeviceSyncPayload(
        version: 1,
        kind: .outgoingMessage,
        logicalMessageID: record.id,
        occurredAt: .now,
        outgoingMessage: DeviceSyncOutgoingMessage(
          recipientUserID: record.payload.recipientUserID,
          recipientUsername: record.payload.recipientUsername,
          body: body,
          sentAt: .now,
          attachment: attachment
        ),
        readState: nil
      ),
      session: session
    )
  }

  func enqueueReadState(peerUsername: String, session: AuthSession) async throws {
    try await enqueueDeviceSync(
      DeviceSyncPayload(
        version: 1,
        kind: .readState,
        logicalMessageID: UUID().uuidString,
        occurredAt: .now,
        outgoingMessage: nil,
        readState: DeviceSyncReadState(peerUsername: peerUsername, readAt: .now)
      ),
      session: session
    )
  }

  func snapshot() async throws -> [Conversation] {
    try await store.conversations()
  }

  private func processIncoming(
    _ message: ServerMessage,
    session: AuthSession,
    selectedConversationID: String?
  ) async throws -> GatewayAcknowledgement? {
    let prepared = try await crypto.prepareDecryption(
      ciphertext: message.ciphertext.data,
      ownerUsername: session.username,
      username: message.senderUsername,
      senderDeviceID: message.senderDeviceID
    )
    do {
      switch try decode(prepared.plaintext) {
      case .message(let body, let attachment):
        let conversationID = message.senderUsername.lowercased()
        try await store.stageIncoming(
          message: message,
          body: body,
          attachment: attachment,
          mutation: prepared.mutation,
          isRead: selectedConversationID == conversationID
        )
      case .deviceSync(let payload):
        guard message.senderUserID == session.userID,
          message.senderUsername.lowercased() == session.username.lowercased()
        else {
          throw MessagingError.invalidPlaintext
        }
        try await store.stageDeviceSyncIncoming(
          message: message,
          payload: payload,
          mutation: prepared.mutation
        )
      }
    } catch {
      await crypto.discard(prepared.mutation)
      throw error
    }
    _ = try await recoverMutations()
    return try await store.acknowledgement(messageID: message.id)
  }

  private func recoverMutations() async throws -> [GatewayAcknowledgement] {
    let mutations = try await store.pendingMutations()
    var acknowledgements: [GatewayAcknowledgement] = []
    for mutation in mutations {
      try await crypto.apply(mutation.mutation)
      try await store.completeMutation(mutation)
      if mutation.direction == .incoming,
        let acknowledgement = try await store.acknowledgement(messageID: mutation.messageID)
      {
        acknowledgements.append(acknowledgement)
      }
    }
    return acknowledgements
  }

  private enum DecodedPayload {
    case message(String, AttachmentCapability?)
    case deviceSync(DeviceSyncPayload)
  }

  private func decode(_ plaintext: Data) throws -> DecodedPayload {
    if let payload = try MessagePayloadCodec.decode(
      plaintext,
      maximumCiphertextSize: maximumAttachmentCiphertextSize
    ) {
      switch payload.kind {
      case .text:
        guard let text = payload.text else {
          throw MessagingError.invalidPlaintext
        }
        return .message(text, nil)
      case .attachment:
        guard let attachment = payload.attachment else {
          throw MessagingError.invalidPlaintext
        }
        return .message(attachment.filename, attachment)
      case .deviceSync:
        guard let deviceSync = payload.deviceSync else {
          throw MessagingError.invalidPlaintext
        }
        return .deviceSync(deviceSync)
      }
    }
    guard let text = String(data: plaintext, encoding: .utf8) else {
      throw MessagingError.invalidPlaintext
    }
    return .message(text, nil)
  }

  private func enqueueDeviceSync(_ payload: DeviceSyncPayload, session: AuthSession) async throws {
    let directory = try await api.ownKeyDirectory(accessToken: session.accessToken)
    guard !directory.devices.isEmpty else {
      return
    }
    let plaintext = try MessagePayloadCodec.encode(.deviceSync(payload))
    let prepared = try await crypto.prepareMessage(
      plaintext: plaintext,
      ownerUsername: session.username,
      username: session.username,
      devices: directory.devices
    )
    let transportID = UUID().uuidString
    do {
      try await store.enqueueDeviceSync(
        transportID: transportID,
        payload: payload,
        plaintext: plaintext,
        recipientUsername: session.username,
        recipientUserID: session.userID,
        envelopes: prepared.envelopes,
        mutation: prepared.mutation,
        sentAt: .now
      )
    } catch {
      await crypto.discard(prepared.mutation)
      throw error
    }
    _ = try await recoverMutations()
  }

  private func unique(_ acknowledgements: [GatewayAcknowledgement])
    -> [GatewayAcknowledgement]
  {
    var values: [String: GatewayAcknowledgement] = [:]
    for acknowledgement in acknowledgements {
      values[acknowledgement.messageID] = acknowledgement
    }
    return values.values.sorted { $0.messageID < $1.messageID }
  }
}
